library/gcm.c

```diff
@@ -42,38 +42,73 @@
 
 /* Used to select the acceleration mechanism */
-#define MBEDTLS_GCM_ACC_SMALLTABLE  0
-#define MBEDTLS_GCM_ACC_LARGETABLE  1
-#define MBEDTLS_GCM_ACC_AESNI       2
-#define MBEDTLS_GCM_ACC_AESCE       3
-
-/*
- * Initialize a context
- */
-void mbedtls_gcm_init(mbedtls_gcm_context *ctx)
-{
-    memset(ctx, 0, sizeof(mbedtls_gcm_context));
-}
+#define MBEDTLS_GCM_ACC_UNSET       0
+#define MBEDTLS_GCM_ACC_SMALLTABLE  1
+#define MBEDTLS_GCM_ACC_LARGETABLE  2
+#define MBEDTLS_GCM_ACC_AESNI       3
+#define MBEDTLS_GCM_ACC_AESCE       4
+
+/* Try to define acceleration as a compile-time constant, if possible, to enable the compiler
+ * to make code-size savings. */
+#if !defined(MBEDTLS_AES_USE_HARDWARE_ONLY)
+    #if defined(MBEDTLS_AESNI_HAVE_CODE) || defined(MBEDTLS_AESCE_HAVE_CODE)
+/* Use run-time detection to select acceleration */
+static uint8_t acceleration = MBEDTLS_GCM_ACC_UNSET;
+        #define MBEDTLS_GCM_RUNTIME_ACCELERATION_DETECTION
+    #else
+/* We know at compile-time it will be software */
+        #if defined(MBEDTLS_GCM_LARGE_TABLE)
+            #define acceleration MBEDTLS_GCM_ACC_LARGETABLE
+        #else
+            #define acceleration MBEDTLS_GCM_ACC_SMALLTABLE
+        #endif
+    #endif
+#else
+/* Either AESCE or AESNI */
+    #if defined(MBEDTLS_AESNI_HAVE_CODE)
+        #define acceleration MBEDTLS_GCM_ACC_AESNI
+    #elif defined(MBEDTLS_AESCE_HAVE_CODE)
+        #define acceleration MBEDTLS_GCM_ACC_AESCE
+    #else
+        #error No AES hardware acceleration available, and MBEDTLS_AES_USE_HARDWARE_ONLY set
+    #endif
+#endif
 
-static inline void gcm_set_acceleration(mbedtls_gcm_context *ctx)
+#if defined(MBEDTLS_GCM_RUNTIME_ACCELERATION_DETECTION)
+static inline void gcm_set_acceleration(void)
 {
+    if (acceleration == MBEDTLS_GCM_ACC_UNSET) {
 #if defined(MBEDTLS_GCM_LARGE_TABLE)
-    ctx->acceleration = MBEDTLS_GCM_ACC_LARGETABLE;
+        acceleration = MBEDTLS_GCM_ACC_LARGETABLE;
 #else
-    ctx->acceleration = MBEDTLS_GCM_ACC_SMALLTABLE;
+        acceleration = MBEDTLS_GCM_ACC_SMALLTABLE;
 #endif
 
 #if defined(MBEDTLS_AESNI_HAVE_CODE)
-    /* With CLMUL support, we need only h, not the rest of the table */
-    if (mbedtls_aesni_has_support(MBEDTLS_AESNI_CLMUL)) {
-        ctx->acceleration = MBEDTLS_GCM_ACC_AESNI;
-    }
+        /* With CLMUL support, we need only h, not the rest of the table */
+        if (mbedtls_aesni_has_support(MBEDTLS_AESNI_CLMUL)) {
+            acceleration = MBEDTLS_GCM_ACC_AESNI;
+        }
 #endif
 
 #if defined(MBEDTLS_AESCE_HAVE_CODE)
-    if (MBEDTLS_AESCE_HAS_SUPPORT()) {
-        ctx->acceleration = MBEDTLS_GCM_ACC_AESCE;
+        if (MBEDTLS_AESCE_HAS_SUPPORT()) {
+            acceleration = MBEDTLS_GCM_ACC_AESCE;
+        }
+#endif
     }
+}
+#endif
+
+/*
+ * Initialize a context
+ */
+void mbedtls_gcm_init(mbedtls_gcm_context *ctx)
+{
+    memset(ctx, 0, sizeof(mbedtls_gcm_context));
+
+#if defined(MBEDTLS_GCM_RUNTIME_ACCELERATION_DETECTION)
```

Comment:
Could be more direct and actually check if `acceleration` is defined

Reply:
True. But I think it's a little less readable.

Reply:
I wasn't thinking about readability but maintainability here. Having a macro that is `#define`d if something else (`acceleration`) is `#define`d seems like a potential for getting the two out-of-sync in future changes